internal/pkg/manifest/transform.go

```diff
@@ -0,0 +1,208 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package manifest
+
+import (
+	"reflect"
+
+	"github.com/imdario/mergo"
+)
+
+var basicKinds = []reflect.Kind{
+	reflect.Bool,
+	reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64,
+	reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64,
+	reflect.Float32, reflect.Float64,
+	reflect.Complex64, reflect.Complex128,
+	reflect.Array, reflect.String, reflect.Slice,
+}
+
+type workloadTransformer struct{}
+
+// Transformer returns custom merge logic for workload's fields.
+func (t workloadTransformer) Transformer(typ reflect.Type) func(dst, src reflect.Value) error {
+	if transform := transformBasic(typ); transform != nil {
+		return transform
+	}
+
+	if typ == reflect.TypeOf(Image{}) {
+		return transformImage()
+	}
+
+	if typ.String() == "map[string]manifest.Volume" {
+		return transformMapStringToVolume()
+	}
+	return nil
+}
+
+type volumeTransformer struct{}
+
+// Transformer returns custom merge logic for volume's fields.
+func (t volumeTransformer) Transformer(typ reflect.Type) func(dst, src reflect.Value) error {
+	return transformBasic(typ)
+}
+
+func transformBasic(typ reflect.Type) func(dst, src reflect.Value) error {
+	if typ.Kind() == reflect.Slice {
+		return transformSlice()
+	}
+
+	if typ.Kind() == reflect.Ptr {
+		for _, k := range basicKinds {
+			if typ.Elem().Kind() == k {
+				return transformPBasic()
+			}
+		}
+
+		if typ.Elem().Kind() == reflect.Struct {
+			return transformPStruct()
+		}
+	}
+	return nil
+}
+
+func transformSlice() func(dst, src reflect.Value) error {
+	return func(dst, src reflect.Value) error {
+		dst.Set(src)
+		return nil
+	}
+}
+
+func transformMapStringToVolume() func(dst, src reflect.Value) error {
+	return func(dst, src reflect.Value) error {
+		for _, key := range src.MapKeys() {
+			srcElement := src.MapIndex(key)
+			if !srcElement.IsValid() {
+				continue
+			}
+			dstElement := dst.MapIndex(key)
```

Comment:
Good call! if `dstElement` is empty, then the key should be just set as `srcElement` without needing to do any merging.

The test case work except that there is place: https://github.com/aws/copilot-cli/blob/4e973535197036cd11bbb26f37d87597fa257750/internal/pkg/manifest/storage_applyenv_test.go#L95, in addition to `EFS`, it also needs `MountPointOpts` because this field is populated in the original manifest. Since for any `Map` we do override-and-append, `MountPointOpts` should be kept. This particular unit test needs to be fixed (not the code).

I don't comment out any of the `FAILED_TEST` because that'd be a bug fix, but this PR meant to be upgrading-and-maintaining. I want to open a new PR for bug fix so that it's clear what's being fixed!

Reply:
Gotcha that makes sense to me!